src/audio/asrc/asrc.c

```diff
@@ -860,8 +860,31 @@ static int asrc_reset(struct processing_module *mod)
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_3
+static int asrc_init_shim(struct processing_module *mod)
+{
+	struct ipc_config_asrc *ipc_asrc = rballoc(0, SOF_MEM_CAPS_RAM, sizeof(*ipc_asrc));
+	struct module_data *mod_data = &mod->priv;
+	const struct sof_ipc_comp_asrc *asrc =
+		(const struct sof_ipc_comp_asrc *)mod_data->cfg.init_data;
+
+	if (!ipc_asrc)
+		return -ENOMEM;
+	// TODO size checks
+	ipc_asrc->source_rate = asrc->source_rate;
+	ipc_asrc->sink_rate = asrc->sink_rate;
+	ipc_asrc->asrc_get_asynchronous_mode = asrc->asrc_verify_params;
+	ipc_asrc->operation_mode = asrc->operation_mode;
+
+	rfree(mod_data->cfg.init_data);
+	mod_data->cfg.init_data = ipc_asrc;
+	mod_data->cfg.data = ipc_asrc;
+	return asrc_init(mod);
+}
```

Comment:
Yeah, this is clean and also super easy in the future to validate the module specific params used at init e.g. rate, format, mode. Everything is all in one place for each module.

Reply:
Sounds good, I will work on moving the rest of the type casting then we can test this.